cluster-api-state-metrics-adoption/README.md

```diff
@@ -0,0 +1,237 @@
+# Cluster API state metrics
+
+<!-- toc -->
+- [Glossary](#glossary)
+- [Problem statement](#problem-statement)
+- [Cluster API state metrics](#cluster-api-state-metrics)
+  - [Infrastructure provider specific metrics](#infrastructure-provider-specific-metrics)
+    - [all providers in one binary](#all-providers-in-one-binary)
+    - [code separation per provider](#code-separation-per-provider)
+    - [Conclusion](#conclusion)
+- [Tasks and Stories](#tasks-and-stories)
+  - [Story 1](#story-1)
+  - [Story 2](#story-2)
+  - [Story 3](#story-3)
+<!-- /toc -->
+
+This RFC motivates the adoption of the existing Cluster API state metrics project into our own workflow.
+
+## Glossary
+
+- Infrastructure providers - Cluster API infrastructure implementation for e.g. AWS (CAPA), OpenStack (CAPO), Google Compute Platorm (CAPG), ...
+- `cluster-api-state-metrics` (CASM)
+- Management Cluster (MC) is the Kubernetes Cluster where CAPI components are deployed and which is responsible for creating new Kubernetes Clusters (WCs)
+- Workload Cluster (WC)
+
+## Problem statement
+
+As Cluster Operator we want to get know the state of any Cluster API related object (no matter if Cluster API Core or infrastructure provider specific).
+
+As different states have impact on the expected behavior of the reconciliation e.g.
+
+- A cluster upgrade won't finish successful if an already existing OpenStackMachine is in an error state (even if the machine is in a transient error).
+- It's possible to set nearly all Cluster API related objects into a `paused` state to skip reconciliation for that - primary done by humans.
+
+All these states might lead to different kind of troubleshooting sessions and possibly issues on customer side.
+
+## Cluster API state metrics
+
+As per design `kube-state-metrics` only takes care of Kubernetes resources and all existing Cluster API controllers don't have object related metrics yet, [`cluster-api-state-metrics`](https://github.com/mercedes-benz/cluster-api-state-metrics) was created by Mercedes-Benz.
+
+The CASM code will be contributed to the Cluster API project (Process can be tracked in [issue #6458](https://github.com/kubernetes-sigs/cluster-api/issues/6458)).
+
+### Infrastructure provider specific metrics
+
+Currently CASM only provide metrics of Cluster API Core components (e.g. `clusters`, `kubeadmcontrolplanes`,`machinedeployments`,`machines` or `machinesets`) and it's currently not clear how infrastructure provider specific metrics can be added/generated once the entire code base is living under `kubernetes-sigs/cluster-api`.
+
+But as the code isn't that complex it's very easy to extend the existing implementation to support infrastructure provider specific metrics.
+
+Beside generated changes in `go.mod` and `go.sum` only the following changes are needed to add the two new metrics `capi_openstackcluster_created` and `capi_openstackcluster_paused` into CASM.
```

Comment:
would it be so complex&dirty to add different providers by config? like passing GVK as config + using unstructured in the code.

Reply:
> would it be so complex&dirty to add different providers by config? like passing GVK as config + using unstructured in the code.

my assumption was to align with the already existing implementation (import the type and reuse it) as much as possible.

If we're using `unstructured` objects, we have to deal with different fields for all used different providers, not sure if it's worth to doing it.

@fiunchinho already mentioned this above:
> As it's mentioned in the RFC, trying to put all potential providers metrics implementations into a single app would bring many many code dependencies into the project (or a complex use of `unstructured` objects using conditionals to use different fields for different providers?)

IMO:
As upstream implementation doesn't have a solution for provider specific metrics yet i guess it's worth to start with a simple solution for now (as the benefits for us might be huge on a short term). And once the discussion starts upstream we could take part there as we already have to refactor this component once upstream has found a good solution. WDYT @erkanerol ?

Reply:
I think it is the simplest solution for us (OpenStack provider) but I am not sure it is the most optimal solution for the whole GiantSwarm. We can mention this in the next kaas sync. 

Nevertheless, I don't see any big risk in starting with this solution and then improving in time. So +1 from my side.

Reply:
added a note to the upcoming KaaS-Sync

Reply:
what was the conclusion in that sync?

Reply:
> what was the conclusion in that sync?

that all others will take a look at this PR. Will propagate it again in the next couple of days.

Reply:
whats left for this rfc now?

Reply:
my time to convert it into a new RFC ... still on my list